From a7d2ca19c5695a923761f1e8c2cf932e77017e0e Mon Sep 17 00:00:00 2001
From: Bhagavathi Perumal S <bperumal@codeaurora.org>
Date: Fri, 13 Oct 2017 19:53:42 +0530
Subject: [PATCH] mac80211: Add offload support to send null func

This allows users to get intant ack signal strength of
last transmitted qos null frame.

Signed-off-by: Bhagavathi Perumal S <bperumal@codeaurora.org>
---
 include/net/mac80211.h    |    4 ++++
 net/mac80211/cfg.c        |    4 ++++
 net/mac80211/debugfs.c    |    1 +
 net/mac80211/driver-ops.h |   16 ++++++++++++++--
 net/mac80211/sta_info.c   |    4 ++++
 net/mac80211/sta_info.h   |    1 +
 net/mac80211/status.c     |    7 ++++++-
 7 files changed, 34 insertions(+), 3 deletions(-)

--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -2110,6 +2110,7 @@ enum ieee80211_hw_flags {
 	IEEE80211_HW_REPORTS_LOW_ACK,
 	IEEE80211_HW_SUPPORTS_TX_FRAG,
 	IEEE80211_HW_SUPPORTS_80211_ENCAP,
+	IEEE80211_HW_SEND_NULL_FUNC,
 
 	/* keep last, obviously */
 	NUM_IEEE80211_HW_FLAGS
@@ -3733,6 +3734,9 @@ struct ieee80211_ops {
 	void (*del_nan_func)(struct ieee80211_hw *hw,
 			    struct ieee80211_vif *vif,
 			    u8 instance_id);
+	int (*send_qos_null_func)(struct ieee80211_hw *hw,
+                                   struct ieee80211_vif *vif,
+                                   struct ieee80211_sta *sta);
 };
 
 /**
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -3401,6 +3401,10 @@ static int ieee80211_probe_client(struct
 		ret = -ENOLINK;
 		goto unlock;
 	}
+	if (ieee80211_hw_check(&local->hw, SEND_NULL_FUNC)) {
+		ret = drv_send_qos_null_func(local, sdata, &sta->sta);
+		goto unlock;
+	}
 
 	if (qos) {
 		fc = cpu_to_le16(IEEE80211_FTYPE_DATA |
--- a/net/mac80211/debugfs.c
+++ b/net/mac80211/debugfs.c
@@ -212,6 +212,7 @@ static const char *hw_flag_names[] = {
 	FLAG(REPORTS_LOW_ACK),
 	FLAG(SUPPORTS_TX_FRAG),
 	FLAG(SUPPORTS_80211_ENCAP),
+	FLAG(SEND_NULL_FUNC),
 #undef FLAG
 };
 
--- a/net/mac80211/driver-ops.h
+++ b/net/mac80211/driver-ops.h
@@ -1248,4 +1248,16 @@ static inline void drv_del_nan_func(stru
 	trace_drv_return_void(local);
 }
 
+static inline int drv_send_qos_null_func(struct ieee80211_local *local,
+					 struct ieee80211_sub_if_data *sdata,
+					 struct ieee80211_sta *sta)
+{
+	int ret = -EOPNOTSUPP;
+
+	if (local->ops->send_qos_null_func)
+		ret = local->ops->send_qos_null_func(&local->hw, &sdata->vif,
+						     sta);
+
+	return ret;
+}
 #endif /* __MAC80211_DRIVER_OPS */
--- a/net/mac80211/sta_info.c
+++ b/net/mac80211/sta_info.c
@@ -2287,6 +2287,10 @@ void sta_set_sinfo(struct sta_info *sta,
 		sinfo->filled |= BIT(NL80211_STA_INFO_EXPECTED_THROUGHPUT);
 		sinfo->expected_throughput = thr;
 	}
+	if (!(sinfo->filled & BIT_ULL(NL80211_STA_INFO_ACK_SIGNAL))) {
+		sinfo->ack_signal = sta->status_stats.last_ack_signal;
+		sinfo->filled |= BIT_ULL(NL80211_STA_INFO_ACK_SIGNAL);
+	}
 }
 
 u32 sta_get_expected_throughput(struct sta_info *sta)
--- a/net/mac80211/sta_info.h
+++ b/net/mac80211/sta_info.h
@@ -544,6 +544,7 @@ struct sta_info {
 		u64 msdu_retries[IEEE80211_NUM_TIDS + 1];
 		u64 msdu_failed[IEEE80211_NUM_TIDS + 1];
 		unsigned long last_ack;
+		s32 last_ack_signal;
 	} status_stats;
 
 	/* Updated from TX path only, no locking requirements */
--- a/net/mac80211/status.c
+++ b/net/mac80211/status.c
@@ -187,9 +187,14 @@ static void ieee80211_frame_acked(struct
 	struct ieee80211_mgmt *mgmt = (void *) skb->data;
 	struct ieee80211_local *local = sta->local;
 	struct ieee80211_sub_if_data *sdata = sta->sdata;
+	struct ieee80211_tx_info *txinfo = IEEE80211_SKB_CB(skb);
 
-	if (ieee80211_hw_check(&local->hw, REPORTS_TX_ACK_STATUS))
+	if (ieee80211_hw_check(&local->hw, REPORTS_TX_ACK_STATUS)) {
 		sta->status_stats.last_ack = jiffies;
+		if (txinfo->status.ack_signal != 0)
+			sta->status_stats.last_ack_signal =
+						txinfo->status.ack_signal;
+	}
 
 	if (ieee80211_is_data_qos(mgmt->frame_control)) {
 		struct ieee80211_hdr *hdr = (void *) skb->data;
@@ -959,6 +964,11 @@ void ieee80211_tx_status_ext(struct ieee
 			/* Track when last TDLS packet was ACKed */
 			if (test_sta_flag(sta, WLAN_STA_TDLS_PEER_AUTH))
 				sta->status_stats.last_tdls_pkt_time = jiffies;
+			if (info->status.ack_signal != 0) {
+				sta->status_stats.last_ack_signal =
+						info->status.ack_signal;
+				return;
+			}
 		} else {
 			ieee80211_lost_packet(sta, info);
 		}
